### PR TITLE
Add conv-tasnet model

### PR DESCRIPTION
**Pull request by mthrok**
See https://github.com/pytorch/audio/issues/897

### CODE REVIEW

File: examples/source_separation/conv_tasnet/model.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        for layer in self.conv_layers:
+            residual, skip = layer(feats)
+            if residual is not None:  # the last conv layer does not produce residual
+                feats = feats + residual
+            output = output + skip
+        output = self.output_layer(output)
+        return output.view(batch_size, self.num_speakers, self.input_dim, -1)
+
+
+class ConvTasNet(torch.nn.Module):
```
Comment (vincentqb) [inline examples/source_separation/conv_tasnet/model.py:125]: Paper reference?

Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:159]: done

File: examples/source_separation/conv_tasnet/model.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
+"""Implements Conv-TasNet with building blocks of it."""
+
+from typing import Tuple, Optional
+
+import torch
+
+
+class ConvBlock(torch.nn.Module):
```
Comment (vincentqb) [inline examples/source_separation/conv_tasnet/model.py:8]: not in [torchaudiomodels](https://github.com/pytorch/audio/tree/master/torchaudio/models)?

File: examples/source_separation/conv_tasnet/model.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        self.decoder = torch.nn.ConvTranspose1d(
+            in_channels=enc_num_feats,
+            out_channels=1,
+            kernel_size=enc_kernel_size,
+            stride=self.enc_stride,
+            padding=self.enc_stride,
+            bias=False,
+        )
+
+    def _pad_input(self, input: torch.Tensor) -> Tuple[torch.Tensor, int]:
```
Comment (vincentqb) [inline examples/source_separation/conv_tasnet/model.py:249]: This seems more general and reusable than just for this model?

Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:249]: I do not think the assumption of `padding == filter_width // 2` to be generic enough.
We can do that kind of refactoring as needed once we have another model that requires the same kind of padding.

File: examples/source_separation/conv_tasnet/model.py (hunk 2 above)
Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:8]: I will handle that in the later PR.

File: examples/source_separation/conv_tasnet/model.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+        kernel_size: int,
+        padding: int,
+        dilation: int = 1,
+        causal: bool = False,
+        no_residual: bool = False,
+    ):
+        super().__init__()
+
+        if causal:
+            raise NotImplementedError("causal=True is not implemented")
```
Comment (vincentqb) [inline examples/source_separation/conv_tasnet/model.py:39]: Why having this in the first place?

File: examples/source_separation/conv_tasnet/model.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        num_sources: int,
+        kernel_size: int,
+        num_feats: int,
+        num_hidden: int,
+        num_layers: int,
+        num_stacks: int,
+        causal: bool = False,
+    ):
+        if causal:
+            raise NotImplementedError("causal=True is not implemented")
```
Comment (vincentqb) [inline examples/source_separation/conv_tasnet/model.py:115]: same

File: examples/source_separation/conv_tasnet/model.py (hunk 3 above)
Comment (vincentqb) [reply examples/source_separation/conv_tasnet/model.py:249]: Is there a way to factor this code as a static-like method then so as to decouple it from the rest?

File: examples/source_separation/conv_tasnet/model.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        msk_num_feats: int = 128,
+        msk_num_hidden_feats: int = 512,
+        msk_num_layers: int = 8,
+        msk_num_stacks: int = 3,
+        causal: bool = False,
+    ):
+        super().__init__()
+
+        if causal:
+            raise NotImplementedError("causal=True is not implemented")
```
Comment (vincentqb) [inline examples/source_separation/conv_tasnet/model.py:216]: same

File: examples/source_separation/conv_tasnet/model.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        input_dim: int,
+        num_sources: int,
+        kernel_size: int,
+        num_feats: int,
+        num_hidden: int,
+        num_layers: int,
+        num_stacks: int,
+        causal: bool = False,
+    ):
+        if causal:
```
Comment (cpuhrsch) [inline examples/source_separation/conv_tasnet/model.py:114]: Do you want to remove the flag and just add this to the docs?

File: examples/source_separation/conv_tasnet/model.py (hunk 4 above)
Comment (vincentqb) [reply examples/source_separation/conv_tasnet/model.py:39]: you can just move this into the doc

File: examples/source_separation/conv_tasnet/model.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+                        padding=multi,
+                        causal=causal,
+                        # The last ConvBlock does not need residual
+                        no_residual=(l == (num_layers - 1) and s == (num_stacks - 1)),
+                    )
+                )
+                self.receptive_field += (
+                    kernel_size if s == 0 and l == 0 else (kernel_size - 1) * multi
+                )
+        self.output_layer = torch.nn.Sequential(
```
Comment (cpuhrsch) [inline examples/source_separation/conv_tasnet/model.py:148]: minor nit: You could turn this into control flow in the forward function instead of using the declarative nn.Sequential container (which encodes a very particular kind of control flow).

File: examples/source_separation/conv_tasnet/model.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+        hidden_channels: int,
+        kernel_size: int,
+        padding: int,
+        dilation: int = 1,
+        causal: bool = False,
+        no_residual: bool = False,
+    ):
+        super().__init__()
+
+        if causal:
```
Comment (cpuhrsch) [inline examples/source_separation/conv_tasnet/model.py:38]: Do you want to remove this flag and add to documentation it's not supported?

File: examples/source_separation/conv_tasnet/model.py (hunk 3 above)
Comment (cpuhrsch) [inline examples/source_separation/conv_tasnet/model.py:249]: I'd change the name of this to avoid confusion (e.g. someone might confuse it with a function such as https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pad_sequence.html )

**Comment by vincentqb** [comment]
Can you also add the tests you mentioned?

File: examples/source_separation/conv_tasnet/model.py (hunk 3 above)
Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:242]: Can you help come up with the name?
Padding is what is curried out in this method, yet it's still difficult to summarize the why the padding is curried out in a few words. Also this is an internal method, so when someone reads this I assume they have some context of the model and they can assume this is related to convolution operation and encoder/decoder operation.

Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:242]: Why static? This method uses the model parameters. It should be tied to the model instance and has a very specific purpose. What is your justification of decoupling it?

File: examples/source_separation/conv_tasnet/model.py (hunk 8 above)
Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:143]: I see your point, and will consider that in the future.

Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:143]: It's just changing this now and I will have to retrain the model before the model release. (well, it might be worth doing it) so I am a bit hesitant to do.

Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:143]: a (logically weak) excuse is that since this is a reference implementation, so we do not intend to change it, but we cannot be sure about the future.

Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py:143]: actually, I realized that there are still some improvements I can apply. I will apply your suggestion and run the training again.

File: examples/source_separation/conv_tasnet/model.py (hunk 3 above)
Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py]: Changed it to `_align_num_frames_with_strides`

File: examples/source_separation/conv_tasnet/model.py (hunk 8 above)
Comment (mthrok) [reply examples/source_separation/conv_tasnet/model.py]: done

**Comment by mthrok** [comment]
Superseded by #920
